File: python/ltx/src/interplanet_ltx/_security.py
"""
_security.py — Cryptographic Identity and Canonical JSON
Story 28.1 — LTX Python SDK security primitives

canonical_json: RFC 8785 / JCS compliant serialisation.
NIK: Node Identity Key — Ed25519 key pair, nodeId derivation, expiry checks.

Uses the Python standard library only (hashlib, secrets) plus the
`cryptography` package for Ed25519 key generation. If `cryptography` is not
available, falls back to `PyNaCl` (nacl). A clear ImportError is raised if
neither is installed.
"""

from __future__ import annotations

import base64
import hashlib
import json
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, Optional


# ── Canonical JSON (RFC 8785 / JCS) ──────────────────────────────────────────


def canonical_json(obj: Any) -> str:
    """
    Produce RFC 8785 / JCS canonical JSON.

    - Object keys are sorted lexicographically (Unicode code-point order).
    - Array element order is preserved.
    - No optional whitespace.

    Returns a str (UTF-8 content — encode to bytes if a bytes payload is needed).
    """
    if obj is None:
        return 'null'
    if isinstance(obj, bool):
        return 'true' if obj else 'false'
    if isinstance(obj, (int, float)):
        return json.dumps(obj)
    if isinstance(obj, str):
        return json.dumps(obj)
    if isinstance(obj, list):
        return '[' + ','.join(canonical_json(v) for v in obj) + ']'
    if isinstance(obj, dict):
        keys = sorted(obj.keys())
        return '{' + ','.join(json.dumps(k) + ':' + canonical_json(obj[k]) for k in keys) + '}'
    raise TypeError(f'canonical_json: unsupported type {type(obj).__name__!r}')


# ── Ed25519 key generation helpers ────────────────────────────────────────────


def _generate_ed25519() -> tuple[bytes, bytes]:
    """
    Generate an Ed25519 key pair.
    Returns (raw_private_seed_32_bytes, raw_public_key_32_bytes).

    Tries `cryptography` first, then `PyNaCl`, raises ImportError if neither.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
        private_key = Ed25519PrivateKey.generate()
        # Raw private seed (32 bytes)
        from cryptography.hazmat.primitives.serialization import (
            Encoding, NoEncryption, PrivateFormat, PublicFormat,
        )
        raw_priv = private_key.private_bytes(
            encoding=Encoding.Raw,
            format=PrivateFormat.Raw,
            encryption_algorithm=NoEncryption(),
        )
        raw_pub = private_key.public_key().public_bytes(
            encoding=Encoding.Raw,
            format=PublicFormat.Raw,
        )
        return raw_priv, raw_pub
    except ImportError:
        pass

    try:
        import nacl.signing  # type: ignore[import]
        signing_key = nacl.signing.SigningKey.generate()
        raw_priv = bytes(signing_key)
        raw_pub  = bytes(signing_key.verify_key)
        return raw_priv, raw_pub
    except ImportError:
        pass

    raise ImportError(
        'NIK key generation requires either the `cryptography` or `PyNaCl` package. '
        'Install one: pip install cryptography'
    )


# ── NIK functions ─────────────────────────────────────────────────────────────


def generate_nik(valid_days: int = 365, node_label: str = '') -> Dict[str, Any]:
    """
    Generate a new Node Identity Key (NIK) record.

    Parameters
    ----------
    valid_days : int
        Validity period in days (default 365).
    node_label : str
        Optional human-readable label stored in the NIK record.

    Returns
    -------
    dict with keys:
        ``nik``            — NIK record dict
        ``private_key_b64`` — base64url-encoded raw private seed (32 bytes)
    """
    raw_priv, raw_pub = _generate_ed25519()

    pub_b64 = base64.urlsafe_b64encode(raw_pub).rstrip(b'=').decode()

    # nodeId: base64url of first 16 bytes of SHA-256(raw public key)
    digest  = hashlib.sha256(raw_pub).digest()
    node_id = base64.urlsafe_b64encode(digest[:16]).rstrip(b'=').decode()

    now         = datetime.now(timezone.utc)
    valid_until = now + timedelta(days=valid_days)

    nik: Dict[str, Any] = {
        'nodeId':     node_id,
        'publicKey':  pub_b64,
        'algorithm':  'Ed25519',
        'validFrom':  now.strftime('%Y-%m-%dT%H:%M:%S.') + f'{now.microsecond // 1000:03d}Z',
        'validUntil': valid_until.strftime('%Y-%m-%dT%H:%M:%S.') + f'{valid_until.microsecond // 1000:03d}Z',
        'keyVersion': 1,
    }
    if node_label:
        nik['label'] = node_label

    priv_b64 = base64.urlsafe_b64encode(raw_priv).rstrip(b'=').decode()

    return {
        'nik':             nik,
        'private_key_b64': priv_b64,
    }


def nik_fingerprint(nik: Dict[str, Any]) -> str:
    """
    Return the SHA-256 hex fingerprint of a NIK's public key.

    Parameters
    ----------
    nik : dict
        A NIK record with a ``publicKey`` field (base64url, 43 chars / 32 bytes).

    Returns
    -------
    str
        64-character lowercase hex string.
    """
    raw_pub = base64.urlsafe_b64decode(nik['publicKey'] + '==')
    return hashlib.sha256(raw_pub).hexdigest()


def is_nik_expired(nik: Dict[str, Any]) -> bool:
    """
    Return True if the NIK's ``validUntil`` timestamp is in the past.

    Parameters
    ----------
    nik : dict
        A NIK record with a ``validUntil`` ISO 8601 UTC string.
    """
    valid_until = datetime.fromisoformat(nik['validUntil'].replace('Z', '+00:00'))
    return datetime.now(timezone.utc) > valid_until


# ── COSE_Sign1 SessionPlan signing ────────────────────────────────────────────


def sign_plan(plan: Dict[str, Any], private_key_b64: str) -> Dict[str, Any]:
    """
    Sign an LTX session plan using a simplified COSE_Sign1-compatible structure.

    Wire format (JSON envelope)::

        {
            "plan": <plan dict>,
            "coseSign1": {
                "protected": "<base64url of canonical JSON of {\"alg\": -19}>",
                "unprotected": {"kid": "<first 8 chars of first node id>"},
                "payload": "<base64url of canonical JSON of plan>",
                "signature": "<base64url of Ed25519 signature over Sig_Structure>"
            }
        }

    The Sig_Structure signed is the canonical JSON (RFC 8785) of::

        ["Signature1", "<protected-b64>", "", "<payload-b64>"]

    Parameters
    ----------
    plan : dict
        LTX plan config (any JSON-serialisable dict).
    private_key_b64 : str
        Base64url-encoded raw 32-byte Ed25519 private seed (from generate_nik).

    Returns
    -------
    dict
        Signed envelope ``{"plan": plan, "coseSign1": {...}}``.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
    except ImportError:
        raise ImportError(
            'sign_plan requires the `cryptography` package. '
            'Install it: pip install cryptography'
        )

    # Decode private seed
    raw_seed = base64.urlsafe_b64decode(private_key_b64 + '==')
    priv_key = Ed25519PrivateKey.from_private_bytes(raw_seed)

    # Protected header: canonical JSON of { alg: -19 } (-19 = EdDSA in COSE)
    protected_header = canonical_json({'alg': -19})
    protected_b64 = base64.urlsafe_b64encode(protected_header.encode()).rstrip(b'=').decode()

    # Payload: canonical JSON of the plan
    payload_str = canonical_json(plan)
    payload_b64 = base64.urlsafe_b64encode(payload_str.encode()).rstrip(b'=').decode()

    # Sig_Structure: canonical JSON of the array
    sig_structure = canonical_json(['Signature1', protected_b64, '', payload_b64])
    sig_bytes = priv_key.sign(sig_structure.encode())
    sig_b64 = base64.urlsafe_b64encode(sig_bytes).rstrip(b'=').decode()

    # Derive NIK nodeId from public key to use as kid.
    # nodeId = base64url of first 16 bytes of SHA-256(raw public key), same as generate_nik.
    from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
    raw_pub_for_kid = priv_key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
    kid_hash = hashlib.sha256(raw_pub_for_kid).digest()
    kid = base64.urlsafe_b64encode(kid_hash[:16]).rstrip(b'=').decode()

    return {
        'plan': plan,
        'coseSign1': {
            'protected': protected_b64,
            'unprotected': {'kid': kid},
            'payload': payload_b64,
            'signature': sig_b64,
        },
    }


def verify_plan(
    signed_envelope: Dict[str, Any],
    key_cache: Dict[str, Any],
) -> Dict[str, Any]:
    """
    Verify a COSE_Sign1-signed session plan envelope.

    Parameters
    ----------
    signed_envelope : dict
        Output from ``sign_plan()`` — a dict with ``plan`` and ``coseSign1`` keys.
    key_cache : dict
        Plain dict mapping nodeId → NIK record.

    Returns
    -------
    dict
        ``{"valid": bool, "reason": str|None}``
        ``reason`` is ``None`` when valid, otherwise a short error code.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        from cryptography.exceptions import InvalidSignature
    except ImportError:
        raise ImportError(
            'verify_plan requires the `cryptography` package. '
            'Install it: pip install cryptography'
        )

    cose = signed_envelope.get('coseSign1')
    if not cose:
        return {'valid': False, 'reason': 'missing_cose_sign1'}

    kid = cose.get('unprotected', {}).get('kid', '')

    # Look up signer's NIK in key_cache (dict nodeId → NIK)
    signer_nik = key_cache.get(kid)
    if not signer_nik:
        signer_nik = next(
            (n for n in key_cache.values() if n.get('nodeId', '').startswith(kid)),
            None,
        )
    if not signer_nik:
        return {'valid': False, 'reason': 'key_not_in_cache'}
    if is_nik_expired(signer_nik):
        return {'valid': False, 'reason': 'key_expired'}

    # Reconstruct Sig_Structure
    sig_structure = canonical_json(['Signature1', cose['protected'], '', cose['payload']])

    # Reconstruct public key from raw 32 bytes
    raw_pub = base64.urlsafe_b64decode(signer_nik['publicKey'] + '==')
    pub_key = Ed25519PublicKey.from_public_bytes(raw_pub)

    # Verify signature
    sig_bytes = base64.urlsafe_b64decode(cose['signature'] + '==')
    try:
        pub_key.verify(sig_bytes, sig_structure.encode())
        valid = True
    except InvalidSignature:
        valid = False

    if not valid:
        return {'valid': False, 'reason': 'signature_invalid'}

    # Also verify that the embedded payload matches the plan
    payload_str = base64.urlsafe_b64decode(cose['payload'] + '==').decode()
    plan = signed_envelope.get('plan', {})
    if payload_str != canonical_json(plan):
        return {'valid': False, 'reason': 'payload_mismatch'}

    return {'valid': True, 'reason': None}


# ── Sequence Tracking ─────────────────────────────────────────────────────────


class SequenceTracker:
    """
    Per-(plan_id, node_id) monotonic sequence tracker for replay protection.

    Tracks both outbound (next_seq) and inbound (record_seq) sequence numbers
    per node_id. Rejects bundles where seq <= last_seen_seq.

    Parameters
    ----------
    plan_id : str
        Plan identifier used to namespace storage keys.
    storage : optional
        Object with ``get(key, default)`` and ``set(key, val)`` methods.
        Defaults to an in-memory dict.
    """

    def __init__(self, plan_id: str, storage=None) -> None:
        self._prefix = f'ltx_seq_{plan_id}_'
        self._mem: Dict[str, int] = {}
        self._storage = storage

    def _get(self, key: str) -> int:
        if self._storage is not None:
            return self._storage.get(key, 0)
        return self._mem.get(key, 0)

    def _set(self, key: str, val: int) -> None:
        if self._storage is not None:
            self._storage.set(key, val)
        else:
            self._mem[key] = val

    def next_seq(self, node_id: str) -> int:
        """Increment and return the next outbound sequence number for this node."""
        key = self._prefix + node_id
        current = self._get(key)
        nxt = current + 1
        self._set(key, nxt)
        return nxt

    def record_seq(self, node_id: str, seq: int) -> Dict[str, Any]:
        """
        Record an inbound sequence number.

        Returns
        -------
        dict with keys: accepted (bool), gap (bool), gap_size (int), reason (str, optional)
        """
        key = self._prefix + node_id + '_rx'
        last = self._get(key)
        if seq <= last:
            return {'accepted': False, 'gap': False, 'gap_size': 0, 'reason': 'replay'}
        gap = seq > last + 1
        gap_size = seq - last - 1 if gap else 0
        self._set(key, seq)
        return {'accepted': True, 'gap': gap, 'gap_size': gap_size}

    def last_seen_seq(self, node_id: str) -> int:
        """Return the last accepted inbound seq for node_id (0 if none seen)."""
        return self._get(self._prefix + node_id + '_rx')

    def current_seq(self, node_id: str) -> int:
        """Return the current outbound seq counter for node_id (0 if none sent)."""
        return self._get(self._prefix + node_id)

    def snapshot(self) -> Dict[str, int]:
        """Export in-memory state snapshot for persistence."""
        return dict(self._mem)


def add_seq(bundle: Dict[str, Any], tracker: SequenceTracker, node_id: str) -> Dict[str, Any]:
    """
    Add a ``seq`` field to a bundle dict using the tracker's next sequence number.

    Parameters
    ----------
    bundle : dict
        Bundle object to stamp.
    tracker : SequenceTracker
        Sequence tracker instance.
    node_id : str
        Sending node ID.

    Returns
    -------
    dict
        New bundle with ``seq`` field added.
    """
    result = dict(bundle)
    result['seq'] = tracker.next_seq(node_id)
    return result


def check_seq(
    bundle: Dict[str, Any],
    tracker: SequenceTracker,
    sender_node_id: str,
) -> Dict[str, Any]:
    """
    Check an incoming bundle's ``seq`` field against the tracker.

    Parameters
    ----------
    bundle : dict
        Incoming bundle (should have a ``seq`` key).
    tracker : SequenceTracker
        Sequence tracker instance.
    sender_node_id : str
        Node ID of the sender.

    Returns
    -------
    dict
        Acceptance result with keys: accepted, gap, gap_size, reason (optional).
    """
    if 'seq' not in bundle:
        return {'accepted': False, 'gap': False, 'gap_size': 0, 'reason': 'missing_seq'}
    return tracker.record_seq(sender_node_id, bundle['seq'])

File: elixir/ltx/lib/interplanet_ltx/security.ex
# security.ex — Epic 29 security cascade for InterplanetLtx (Elixir)
# Stories 29.1, 29.4, 29.5
# Uses Erlang :crypto for Ed25519 (OTP 22+)

defmodule InterplanetLtx.Security do
  @moduledoc """
  NIK generation, plan signing/verification, and sequence tracking.
  Uses Erlang :crypto for Ed25519 (OTP 22+).
  """

  # ── Canonical JSON ─────────────────────────────────────────────────────────

  @doc "Produce a deterministic, sorted-key JSON string."
  def canonical_json(value) when is_map(value) do
    sorted = value |> Enum.sort_by(fn {k, _} -> k end)
    inner =
      sorted
      |> Enum.map(fn {k, v} ->
        json_str(to_string(k)) <> ":" <> canonical_json(v)
      end)
      |> Enum.join(",")
    "{" <> inner <> "}"
  end

  def canonical_json(value) when is_list(value) do
    "[" <> (value |> Enum.map(&canonical_json/1) |> Enum.join(",")) <> "]"
  end

  def canonical_json(true),  do: "true"
  def canonical_json(false), do: "false"
  def canonical_json(nil),   do: "null"
  def canonical_json(v) when is_integer(v), do: Integer.to_string(v)
  def canonical_json(v) when is_float(v),
    do: :erlang.float_to_binary(v, [:compact, {:decimals, 15}])
  def canonical_json(v) when is_binary(v), do: json_str(v)
  def canonical_json(v) when is_atom(v),   do: json_str(Atom.to_string(v))

  defp json_str(s) do
    inner =
      s
      |> String.to_charlist()
      |> Enum.map_join(fn
        ?"  -> "\\\""
        ?\\ -> "\\\\"
        ?\n -> "\\n"
        ?\r -> "\\r"
        ?\t -> "\\t"
        c when c < 0x20 ->
          :io_lib.format("\\u~4.16.0b", [c]) |> IO.chardata_to_string()
        c -> <<c::utf8>>
      end)
    "\"" <> inner <> "\""
  end

  # ── generate_nik ───────────────────────────────────────────────────────────

  @doc """
  Generate a new Node Identity Key (NIK).
  Options: valid_days (default 365), node_label (default "").
  Returns %{nik: map, private_key_b64: string}.
  """
  def generate_nik(opts \\ []) do
    valid_days = Keyword.get(opts, :valid_days, 365)
    node_label = Keyword.get(opts, :node_label, "")

    # :crypto.generate_key(:eddsa, :ed25519) returns {pub_32_bytes, priv_32_bytes}
    {pub_raw, priv_raw} = :crypto.generate_key(:eddsa, :ed25519)

    hash    = :crypto.hash(:sha256, pub_raw)
    node_id = hash |> binary_part(0, 16) |> b64u()

    now_ms   = :os.system_time(:millisecond)
    until_ms = now_ms + valid_days * 86_400_000

    nik = %{
      "nodeId"     => node_id,
      "publicKey"  => b64u(pub_raw),
      "algorithm"  => "Ed25519",
      "validFrom"  => ms_to_iso(now_ms),
      "validUntil" => ms_to_iso(until_ms),
      "keyVersion" => 1
    }
    nik = if node_label != "", do: Map.put(nik, "label", node_label), else: nik

    # Store public key alongside private key so we can derive kid later without re-deriving
    %{nik: nik, private_key_b64: b64u(priv_raw), _pub_raw: pub_raw}
  end

  # ── is_nik_expired ─────────────────────────────────────────────────────────

  @doc "Returns true if the NIK validUntil is in the past."
  def is_nik_expired(%{"validUntil" => valid_until}) do
    :os.system_time(:millisecond) > parse_iso_ms(valid_until)
  end
  def is_nik_expired(_), do: true

  # ── sign_plan ──────────────────────────────────────────────────────────────

  @doc """
  Sign an LTX plan.
  private_key_b64 is base64url of the raw 32-byte Ed25519 private key.
  pub_key_b64 (optional) is the matching public key — needed to derive kid.
  Returns %{plan: plan, coseSign1: map}.
  """
  def sign_plan(plan, private_key_b64, pub_key_b64 \\ nil) do
    protected_b64 = b64u(canonical_json(%{"alg" => -19}))
    payload_b64   = b64u(canonical_json(normalise(plan)))

    msg      = canonical_json(["Signature1", protected_b64, "", payload_b64])
               |> :erlang.iolist_to_binary()
    priv_raw = unb64u(private_key_b64)
    sig      = :crypto.sign(:eddsa, :none, msg, [priv_raw, :ed25519])

    kid =
      if pub_key_b64 do
        unb64u(pub_key_b64)
        |> (&:crypto.hash(:sha256, &1)).()
        |> binary_part(0, 16)
        |> b64u()
      else
        # Cannot derive kid without pub key — use a placeholder
        # This case should be avoided; callers should pass the pub key
        b64u(:crypto.hash(:sha256, priv_raw) |> binary_part(0, 16))
      end

    %{
      plan: plan,
      coseSign1: %{
        "protected"   => protected_b64,
        "unprotected" => %{"kid" => kid},
        "payload"     => payload_b64,
        "signature"   => b64u(sig)
      }
    }
  end

  # ── verify_plan ────────────────────────────────────────────────────────────

  @doc "Verify a COSE_Sign1-signed plan envelope."
  def verify_plan(%{coseSign1: cs, plan: plan}, key_cache), do: do_verify(cs, plan, key_cache)
  def verify_plan(%{"coseSign1" => cs, "plan" => plan}, key_cache), do: do_verify(cs, plan, key_cache)

  defp do_verify(cs, plan, key_cache) do
    kid        = (cs["unprotected"] || %{})["kid"]
    signer_nik = find_nik(key_cache, kid)

    cond do
      is_nil(signer_nik) ->
        %{valid: false, reason: "key_not_in_cache"}
      is_nik_expired(signer_nik) ->
        %{valid: false, reason: "key_expired"}
      true ->
        protected = cs["protected"]
        payload   = cs["payload"]
        sig_b64   = cs["signature"]

        msg       = canonical_json(["Signature1", protected, "", payload])
                    |> :erlang.iolist_to_binary()
        sig_bytes = unb64u(sig_b64)
        pub_raw   = unb64u(signer_nik["publicKey"])
        valid     = :crypto.verify(:eddsa, :none, msg, sig_bytes, [pub_raw, :ed25519])

        cond do
          not valid ->
            %{valid: false, reason: "signature_invalid"}
          unb64u(payload) != canonical_json(normalise(plan)) ->
            %{valid: false, reason: "payload_mismatch"}
          true ->
            %{valid: true}
        end
    end
  end

  defp find_nik(cache, kid) when is_map(cache) do
    Map.get(cache, kid) ||
      Enum.find_value(cache, fn {_k, v} ->
        if (v["nodeId"] || v[:nodeId]) == kid, do: v, else: nil
      end)
  end
  defp find_nik(_, _), do: nil

  # ── SequenceTracker ────────────────────────────────────────────────────────

  @doc "Create a new in-memory sequence tracker."
  def new_sequence_tracker(plan_id) do
    {:ok, pid} = Agent.start_link(fn -> %{} end)
    %{plan_id: plan_id, pid: pid, prefix: "ltx_seq_#{plan_id}_"}
  end

  @doc "Get next outbound sequence number for node_id."
  def next_seq(%{pid: pid, prefix: prefix}, node_id) do
    key = prefix <> node_id
    Agent.get_and_update(pid, fn store ->
      n = Map.get(store, key, 0) + 1
      {n, Map.put(store, key, n)}
    end)
  end

  @doc "Add :seq field to a bundle map."
  def add_seq(bundle, tracker, node_id) do
    Map.put(bundle, :seq, next_seq(tracker, node_id))
  end

  @doc "Check incoming seq from sender."
  def check_seq(bundle, tracker, sender_node_id) do
    seq = Map.get(bundle, :seq) || Map.get(bundle, "seq")
    if is_nil(seq) or not is_integer(seq) do
      %{accepted: false, reason: "missing_seq", gap: false, gap_size: 0}
    else
      record_seq(tracker, sender_node_id, seq)
    end
  end

  defp record_seq(%{pid: pid, prefix: prefix}, node_id, seq) do
    key = prefix <> node_id <> "_rx"
    Agent.get_and_update(pid, fn store ->
      last = Map.get(store, key, 0)
      {result, new_store} =
        cond do
          seq <= last ->
            {%{accepted: false, reason: "replay", gap: false, gap_size: 0}, store}
          seq == last + 1 ->
            {%{accepted: true, reason: nil, gap: false, gap_size: 0}, Map.put(store, key, seq)}
          true ->
            gs = seq - last - 1
            {%{accepted: true, reason: nil, gap: true, gap_size: gs}, Map.put(store, key, seq)}
        end
      {result, new_store}
    end)
  end

  # ── Private helpers ────────────────────────────────────────────────────────

  defp b64u(bytes) when is_binary(bytes), do: Base.url_encode64(bytes, padding: false)
  defp b64u(str)  when is_list(str),      do: str |> IO.chardata_to_string() |> b64u()
  defp unb64u(s), do: Base.url_decode64!(s, padding: false)

  defp normalise(v) when is_struct(v) do
    v |> Map.from_struct()
    |> Enum.reduce(%{}, fn {k, val}, acc ->
      Map.put(acc, Atom.to_string(k), normalise(val))
    end)
  end
  defp normalise(v) when is_list(v),  do: Enum.map(v, &normalise/1)
  defp normalise(v) when is_atom(v) and v not in [true, false, nil], do: Atom.to_string(v)
  defp normalise(v), do: v

  defp ms_to_iso(ms) do
    secs   = div(ms, 1000)
    ms_rem = rem(ms, 1000)
    {{y, mo, d}, {h, mi, s}} =
      :calendar.gregorian_seconds_to_datetime(secs + 62_167_219_200)
    :io_lib.format(
      "~4..0B-~2..0B-~2..0BT~2..0B:~2..0B:~2..0B.~3..0BZ",
      [y, mo, d, h, mi, s, ms_rem]
    ) |> IO.chardata_to_string()
  end

  defp parse_iso_ms(nil), do: 0
  defp parse_iso_ms(""),  do: 0
  defp parse_iso_ms(str) when is_binary(str) do
    case Regex.run(
      ~r/^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d+))?Z?$/,
      str
    ) do
      [_, y, mo, d, h, mi, s | rest] ->
        ms_s = case rest do
          [ms] -> ms |> String.slice(0, 3) |> String.pad_trailing(3, "0")
          _    -> "000"
        end
        dt = {
          {String.to_integer(y), String.to_integer(mo), String.to_integer(d)},
          {String.to_integer(h), String.to_integer(mi), String.to_integer(s)}
        }
        gsecs = :calendar.datetime_to_gregorian_seconds(dt) - 62_167_219_200
        gsecs * 1000 + String.to_integer(ms_s)
      _ -> 0
    end
  end
end
